src/hotspot/share/nmt/vmatree.hpp

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2024, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_NMT_VMATREE_HPP
+#define SHARE_NMT_VMATREE_HPP
+
+#include "memory/resourceArea.hpp"
+#include "nmt/nmtNativeCallStackStorage.hpp"
+#include "nmt/nmtTreap.hpp"
+#include "runtime/os.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/growableArray.hpp"
+
+// A VMATree stores a sequence of points on the natural number line.
+// Each of these points stores information about a state change.
+// For example, the state may go from released memory to committed memory,
+// or from committed memory of a certain MEMFLAGS to committed memory of a different MEMFLAGS.
+// The set of points is stored in a balanced binary tree for efficient querying and updating.
+class VMATree {
+  friend class VMATreeTest;
+  // A position in memory.
+  using position = size_t;
+
+  class AddressComparator {
+  public:
+    static int cmp(position a, position b) {
+      if (a < b) return -1;
+      if (a == b) return 0;
+      if (a > b) return 1;
+      ShouldNotReachHere();
+    }
+  };
+
+public:
+  enum class StateType : uint8_t { Reserved, Committed, Released };
+
+  // Each point has some stack and a flag associated with it.
+  struct Metadata {
+    const NativeCallStackStorage::StackIndex stack_idx;
+    const MEMFLAGS flag;
+
+    Metadata() : stack_idx(), flag(mtNone) {}
+
+    Metadata(NativeCallStackStorage::StackIndex stack_idx, MEMFLAGS flag)
+    : stack_idx(stack_idx), flag(flag) {}
```

Comment:
We don't have the state here, do you mean in `IntervalState` ctr? I'm doing `!(type == Released) || flag == mtNone` instead of `!=` to make the usage of `~P v Q equiv. P => Q` obvious.

Reply:
Sure, works for me